src/meta-srv/src/selector/weight_based.rs

```diff
@@ -0,0 +1,222 @@
+///
+/// We can build another selector based on this selector.
+/// For example: [crate::selector::load_based::LoadBasedSelector].
+pub struct WeightedBasedSelector<W, C> {
+    weighted_choose: RwLock<W>,
+    weight_compute: C,
+}
+
+impl<W, C> WeightedBasedSelector<W, C> {
+    pub fn new(weighted_choose: W, weight_compute: C) -> WeightedBasedSelector<W, C> {
+        Self {
+            weighted_choose: RwLock::new(weighted_choose),
+            weight_compute,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl<W, C> Selector for WeightedBasedSelector<W, C>
+where
+    W: WeightedChoose<Peer>,
+    C: WeightCompute,
+{
+    type Context = SelectorContext;
+    type Output = Vec<Peer>;
+
+    async fn select(
+        &self,
+        ns: Namespace,
+        ctx: &Self::Context,
+        opts: SelectorOptions,
+    ) -> Result<Self::Output> {
+        // 1. get alive datanodes.
+        let lease_kvs =
+            lease::alive_datanodes(ns, &ctx.meta_peer_client, ctx.datanode_lease_secs).await?;
+
+        if lease_kvs.is_empty() {
+            return Ok(vec![]);
+        }
+
+        // 2. get stat kvs and filter out expired datanodes.
+        let stat_keys = lease_kvs.keys().map(|k| k.into()).collect();
+        let stat_kvs = filter_out_expired_datanode(
+            ctx.meta_peer_client.get_dn_stat_kvs(stat_keys).await?,
+            &lease_kvs,
+        );
+
+        // 3. make sure the regions of a table are distributed on different datanodes.
+        let stat_kvs = if let (Some(catalog), Some(schema), Some(table)) =
+            (&ctx.table, &ctx.schema, &ctx.table)
+        {
+            filter_out_datanode_by_table(stat_kvs, catalog, schema, table)
+        } else {
+            stat_kvs
+        };
+
+        // 4. check whether the data meets the SelectorOptions.
+        check(&stat_kvs, &opts)?;
+
+        // 5. compute weight array
+        let weight_array = self.weight_compute.compute(&stat_kvs);
+
+        // 6. choose peers from weight_array
+        let mut weighted_choose = self.weighted_choose.write();
+        let selected = choose_peers(weight_array, &opts, &mut *weighted_choose)?;
+
+        Ok(selected)
+    }
+}
+
+pub fn filter_out_expired_datanode(
+    mut stat_kvs: HashMap<StatKey, StatValue>,
+    lease_kvs: &HashMap<LeaseKey, LeaseValue>,
+) -> HashMap<StatKey, StatValue> {
+    lease_kvs
+        .iter()
+        .filter_map(|(lease_k, _)| stat_kvs.remove_entry(&lease_k.into()))
+        .collect()
+}
+
+pub fn filter_out_datanode_by_table(
+    stat_kvs: HashMap<StatKey, StatValue>,
+    catalog: &str,
+    schema: &str,
+    table: &str,
+) -> HashMap<StatKey, StatValue> {
+    stat_kvs
+        .into_iter()
+        .filter(|(_, stat_v)| contains_table(stat_v, catalog, schema, table) != Some(true))
+        .collect()
+}
+
+// This function checks whether the stat_kvs meets the SelectorOptions.
+//
+// If opts.allow_duplication is true, then the stat_kvs must not be empty.
```

Comment:
This comment simply describe the behavior of codes. That's fine. But what's more valuable is the rational behind it.  Why "opts.allow_duplication is true" requires "the stat_kvs must not be empty"?